tamr_toolbox/dataset/manage.py

```diff
@@ -0,0 +1,527 @@
+
+    Args:
+        dataset: An existing TUC dataset
+        attribute_types: dictionary for non-default types, attribute name is the key and
+            AttributeType is the value
+        attribute_descriptions: dictionary for attribute descriptions, attribute name is the
+            key and the attribute description is the value
+        override_existing_types: bool flag, when true will alter existing attributes
+
+    Returns:
+        Updated Dataset
+
+    Raises:
+        requests.HTTPError: If any HTTP error is encountered
+        ValueError: If the dataset is not a source dataset
+        ValueError: If a passed attribute does not exist in the dataset
+        ValueError: If a passed attribute is a primary key and can't be removed
+        ValueError: If there are no updates to attributes in attribute_types or
+            attribute_descriptions arguments
+    """
+    dataset_name = dataset.name
+    if dataset.upstream_datasets():
+        raise ValueError(f"{dataset_name} is not a source dataset")
+
+    # Check description or type changes are passed in
+    if attribute_types is None and attribute_descriptions is None:
+        raise ValueError(
+            """Updates to attributes must be passed in via attribute_types
+            or attribute_descriptions arguments"""
+        )
+
+    # Get list of attributes that need updating from attribute_types and
+    # attribute_descriptions dictionaries
+    attributes = {attr for attr in attribute_types or list()} | {
+        attr for attr in attribute_descriptions or list()
+    }
+
+    # Get current dataset attributes
+    target_dataset_attributes = dataset.attributes
+    target_attribute_dict = {}
+    for attr in target_dataset_attributes.stream():
+        target_attribute_dict[attr.name] = attr
+    existing_attributes = target_attribute_dict.keys()
+    primary_keys = dataset.spec().to_dict()["keyAttributeNames"]
+
+    # Check that all of the attribute names already exist in dataset
+    for attribute_name in attributes:
+        if attribute_name not in existing_attributes:
+            # This attribute does not exist
+            raise ValueError(
+                f"An attribute with name '{attribute_name}' does not exist in {dataset_name}"
+            )
+        elif attribute_name in primary_keys:
+            # Can not edit a primary key
+            raise ValueError(
+                f"The attribute '{attribute_name}' is a primary key and can't be updated"
+            )
+
+    # Update attributes in dataset
+    for attribute_name in attributes:
+        attr_spec_dict = _make_spec_dict(
+            attribute_name=attribute_name,
+            attribute_types=attribute_types,
+            attribute_descriptions=attribute_descriptions,
+        )
+        existing_attribute_spec = target_attribute_dict[attribute_name].spec()
+        new_type_class = attribute_type.from_json(attr_spec_dict["type"])
```

Comment:
**This is the important comment!**
I still think the better behavior here is:
If the attribute type is not specified in the `attribute_types` dictionary (or that parameter is `None`) then the attribute should inherit it's old type, rather than it being converted to the DEFAULT type.  Otherwise I think users might be frustrated that this code keeps trying to change their GEO or other non-DEFAULT attributes' types when they are just trying to update a description.
You could do that with
```suggestion
        if attribute_types is None or attribute_name not in attribute_types:
            new_type_class = attribute_type.from_json(existing_attribute_spec.to_dict()["type"])
        else:
            new_type_class = attribute_type.from_json(attr_spec_dict["type"])
```

Reply:
Make sure to update the function docstring after making this change.

Reply:
Done. Thanks!